### DISCUSSION TITLE
How to update the accessorFn?

### DISCUSSION THREAD

**Discussion by kazukinagata** [TanStack/table#5216] [Q&A]
I am creating a generic data table.
The table data is provided by the API as JSON, and all field types are strings.

The fields that should be treated as numbers are specified by the user, and if specified, I need to cast the value to a number in the accessorFn for sorting, so I am creating a function called makeColumns to dynamically generate columns.

```typescript
function makeColumns(options: Option[]) {
  const columnHelper = createColumnHelper<TableRowData>();
  return options.map((o) =>
    columnHelper.accessor(
      (row) => {
        return o.format === "currency" ? Number(row.visits) : row.visits;
      },
      {
        id: o.key,
        cell: (info) => {
          const value = info.getValue();
          return typeof value === "number"
            ? `$${info.getValue()}`
            : value
        },
      },
    ),
  );
}

const [options, setOptions] = useState<Option[]>([
  { key: "id", format: "none" },
  { key: "visits", format: "currency" },
  { key: "firstName", format: "none" },
]);
const columns = useMemo(() => makeColumns(options), [options]);
```

The problem is that when the format options are changed, the value returned by info.getValue() in the cell callback is not updated.

After a little digging, it appears that the cell callback is called on re-rendering, but the updated accessorFn is not being executed at that time.

Is it possible to update the accessorFn after the table is mounted?

https://codesandbox.io/p/devbox/crazy-swirles-49k3tw?file=%2Fsrc%2Fmain.tsx%3A21%2C4

**Comment by TonyVeigel** [comment]
create a custom filter

**Reply by kazukinagata** [reply]
Thank you for your response. However, it doesn't quite address my original question. My question is whether it's possible to update the accessorFn after the table has been mounted. Additionally, I would like to know if using accessorFn for the purpose of casting data goes against its intended use.

**Reply by ljmerza** [reply]
did you ever figure this out?

**Comment by AdmiJW** [comment]
This issue is also similar as discussed in https://github.com/TanStack/table/discussions/4868. I have found a workaround using `useMemo` which I have also provided in said discussion.

My finding is that, as long as the provided `data` is not updated, nor will the accessor function be updated. Wrapping the `data` in a `useMemo`, returning a different copy of the original `data` array, and including the `options` in the dependency array seems to do the job. Something like:

```ts
const rows = React.useMemo(() => [...data], [data, options]);
```

However for large tables, this may have performance impact, as mentioned in the [docs](https://tanstack.com/table/latest/docs/api/core/table#data)

> Make sure your `data` option is only changing when you want the table to reprocess. Providing an inline `[]` or constructing the `data` array as a new object every time you want to render the table will result in a lot of unnecessary re-processing. This can easily go unnoticed in smaller tables, but you will likely notice it in larger tables.
